cli/lsp/documents.rs

```diff
@@ -733,38 +801,54 @@ impl DocumentsInner {
     version: i32,
     changes: Vec<lsp::TextDocumentContentChangeEvent>,
   ) -> Result<Document, AnyError> {
-    let doc = self.docs.get(specifier).map_or_else(
-      || {
-        Err(custom_error(
-          "NotFound",
-          format!("The specifier \"{}\" was not found.", specifier),
-        ))
-      },
-      Ok,
-    )?;
+    let doc = self
+      .open_docs
+      .get(specifier)
+      .cloned()
+      .or_else(|| {
+        let mut file_system_docs = self.file_system_docs.lock();
+        file_system_docs.docs.remove(specifier)
+      })
+      .map_or_else(
+        || {
+          Err(custom_error(
+            "NotFound",
+            format!("The specifier \"{}\" was not found.", specifier),
+          ))
+        },
+        Ok,
+      )?;
     self.dirty = true;
     let doc = doc.with_change(version, changes, self.get_maybe_resolver())?;
-    self.docs.insert(doc.specifier().clone(), doc.clone());
+    self.open_docs.insert(doc.specifier().clone(), doc.clone());
     Ok(doc)
   }
 
   fn close(&mut self, specifier: &ModuleSpecifier) -> Result<(), AnyError> {
-    let doc = self.docs.get_mut(specifier).map_or_else(
-      || {
-        Err(custom_error(
-          "NotFound",
-          format!("The specifier \"{}\" was not found.", specifier),
-        ))
-      },
-      Ok,
-    )?;
-    *doc = doc.with_closed();
+    let mut file_system_docs = self.file_system_docs.lock();
+    let doc = self
```

Comment:
I am not convinced 100% of this.  We should just lazily `::new()` the local file when needed.  The file could easily be modified by another process outside of the editor, so we can't really trust anything we copy over.

Reply:
Before, it simply would remove the version and clear the nav tree.  I guess why not drop it?  If the file is closed and isn't referenced anywhere else it doesn't need to be in the fs files anyways, and if it does, it will get added.

Reply:
I investigated this in more detail and it looks like `get_workspace_completions` is the only place that uses `.documents(false, _)` where it matters. It seems like this would cause a behaviour change, but I'm not sure. Wouldn't it be possible for someone to have a remote file in the cache, open it, close it, and now it doesn't show up in the completions? Is that a big deal?

Reply:
very very very very unlikely, because if the module is elsewhere in the dependency graph in tsc it would be re-opened as an fs file before that code ever ran, and that code is related to completing specifiers when someone is typing a new import specifier (not referencing workspace symbols for example).  So they would have to be super fast, basically close the file and already be typing a specifier in another open editor file and it likely would "catch-up" after they type the first character.

Reply:
Ok cool. Thanks! I'll make this change.

Reply:
Done in a5c8789c8fabb009fbc19da9800287ecc0d106bf. I'll test this PR out more tomorrow morning and then merge it then.